src/lib.rs

```diff
@@ -1582,9 +1614,27 @@ pub struct ShowMessageRequestParams {
 }
 
 #[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
 pub struct MessageActionItem {
     /// A short title like 'Retry', 'Open Log' etc.
     pub title: String,
+
+    /// Additional attributes that the client preserves and
+    /// sends back to the server. This depends on the client
+    /// capability window.messageActionItem.additionalPropertiesSupport
+    #[cfg(feature = "proposed")]
+    #[serde(flatten)]
+    pub properties: HashMap<String, MessageActionItemProperty>,
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
+#[cfg(feature = "proposed")]
+#[serde(untagged)]
+pub enum MessageActionItemProperty {
```

Comment:
Isn't this just `serde_json::Value`?

Reply:
`serde_json::Value` includes `null` and `arrays`

Reply:
Hm, but this includes free-form value... What happens on the client if we submit `{"a":[], "n": null}`?

Reply:
Here's the typescript def: https://github.com/microsoft/vscode-languageserver-node/commit/d2f29b5737be4dfc375c3824757d72fd7e39366f#diff-e4f1735a759d48800f4f06207efc2e6b9003629d3d39b188ab75bcd98806a094R1016

Reply:
Hm, I wonder why this isn't any, like any other `data` property?

Reply:
@dbaeumer: Can you clarify this?

Reply:
I thought it makes more sense to allow more than one property as we do in formatting options. The data property usually is some from of key. I am open to change this to a data property as well.

Reply:
@matklad I'm inclined to leave this as is for now since it's in the proposal state anyway.

Reply:
sgtm